kernel/work.c

```diff
@@ -0,0 +1,1030 @@
+ *
+ * This may make the work queue thread ready, but as the lock is held it
+ * will not be a reschedule point.  Callers should yield after the lock is
+ * released where appropriate (generally if this returns true).
+ *
+ * @param queue to be notified.  If this is null no notification is required.
+ *
+ * @return true iff the queue was notified and woken, i.e. a reschedule is
+ * pending.
+ */
+static inline bool notify_queue_locked(struct k_work_q *queue)
+{
+	bool rv = false;
+
+	if (queue != NULL) {
+		rv = z_sched_wake(&queue->notifyq, 0, NULL);
+	}
+
+	return rv;
+}
+
+/* Submit an work item to a queue if queue state allows new work.
+ *
+ * Submission is rejected if no queue is provided, or if the queue is
+ * draining and the work isn't being submitted from the queue's
+ * thread (chained submission).
+ *
+ * Invoked with work lock held.
+ * Conditionally notifies queue.
+ *
+ * @param queue the queue to which work should be submitted.  This may
+ * be null, in which case the submission will fail.
+ *
+ * @param work to be submitted
+ *
+ * @retval 1 if successfully queued
+ * @retval -EINVAL if no queue is provided
+ * @retval -ENODEV if the queue is not started
+ * @retval -EBUSY if the submission was rejected (draining, plugged)
+ */
+static inline int queue_submit_locked(struct k_work_q *queue,
+				      struct k_work *work)
+{
+	if (queue == NULL) {
+		return -EINVAL;
+	}
+
+	int ret = -EBUSY;
+	bool chained = (_current == &queue->thread) && !k_is_in_isr();
+	bool draining = flag_test(&queue->flags, K_WORK_QUEUE_DRAIN_BIT);
+	bool plugged = flag_test(&queue->flags, K_WORK_QUEUE_PLUGGED_BIT);
+
+	/* Test for acceptability, in priority order:
+	 *
+	 * * -ENODEV if the queue isn't running.
+	 * * -EBUSY if draining and not chained
+	 * * -EBUSY if plugged and not draining
+	 * * otherwise OK
+	 */
+	if (!flag_test(&queue->flags, K_WORK_QUEUE_STARTED_BIT)) {
+		ret = -ENODEV;
+	} else if (draining && !chained) {
+		ret = -EBUSY;
+	} else if (plugged && !draining) {
+		ret = -EBUSY;
+	} else {
+		sys_slist_append(&queue->pending, &work->node);
+		ret = 1;
+		(void)notify_queue_locked(queue);
+	}
+
+	return ret;
+}
+
+/* Attempt to submit work to a queue.
+ *
+ * The submission can fail if:
+ * * the work is cancelling,
+ * * no candidate queue can be identified;
+ * * the candidate queue rejects the submission.
+ *
+ * Invoked with work lock held.
+ * Conditionally notifies queue.
+ *
+ * @param work the work structure to be submitted
+
+ * @param queuep pointer to a queue reference.  On input this should
+ * dereference to the proposed queue (which may be null); after completion it
+ * will be null if the work was not submitted or if submitted will reference
+ * the queue it was submitted to.  That may or may not be the queue provided
+ * on input.
+ *
+ * @retval 0 if already queued
+ * @retval 1 if unqueued and not running
+ * @retval 2 if unqueued and running
+ * @retval -EBUSY if canceling or submission was rejected by queue
```

Comment:
Possibly.  It's not needed internally, so to be useful it'd have to propagate up to `k_work_submit_to_queue()`, where it would allow distinguishing cancelling from queue is draining or plugged.

Generally if a module chooses to use the cancel feature it's the responsibility of that module to not make use of the work item until the cancellation has resolved (completed or not necessary).  So attempting to submit an item after attempting to cancel it without waiting for the cancellation to complete is rather like the pervasive misuse of the existing API where the cancellation result isn't even checked.

How badly do you want this?

Reply:
We can go without it. I was thinking about a case when one would like (for some reason) to know if interrupt is racing with a timeout. I would add it if it's cheap to do as it would be pain to do it later (in case it would be useful for somebody).

If you think it takes too much effort, please ignore the comment. I have given a green tick already :)